index_release.go

```diff
@@ -19,6 +21,155 @@ type IndexRelease struct {
+		if err != nil {
+			logger.Log("level", "warning", "message", fmt.Sprintf("failed building new release from %s", ir.Version), "stack", fmt.Sprintf("%#v", err))
+			continue
+		}
+
+		release.active = ir.Active
+		release.deprecated = false
+		release.timestamp = ir.Date.Format(indexReleaseTimestampFormat)
+		release.version = ir.Version
+
+		releases = append(releases, release)
+	}
+
+	return releases, nil
+}
+
+func groupBundlesForIndexRelease(ir IndexRelease, bundles map[string]Bundle) ([]Bundle, error) {
+	var groupedBundles []Bundle
+	for _, a := range ir.Authorities {
+		b, found := bundles[a.BundleID()]
+		if !found {
+			return nil, microerror.Maskf(bundleNotFoundError, "IndexRelease v%s contains Authority with bundle ID %s that cannot be found from collected version bundles.", ir.Version, a.BundleID())
+		}
+		groupedBundles = append(groupedBundles, b)
+	}
+
+	return groupedBundles, nil
+}
+
+// deduplicateReleaseChangelog removes duplicate changelog entries in
+// consequtive release entries. Core concept of algorithm here is to first sort
+// releases by their release version and then iterate them and compare current
+// release to previous one that fulfills following requirements: smaller
+// version number and earlier timestamp. Comparison of earlier timestamp is
+// crucial here in order to calculate changelog correctly when newer patch
+// releases have been introduced with lower version number
+// (e.g. [1.0.0, 2.0.0] -> [1.0.0, 1.0.1, 2.0.0, 2.0.1]).
+func deduplicateReleaseChangelog(releases []Release) []Release {
+	if len(releases) < 2 {
+		return releases
+	}
+
+	sort.Sort(SortReleasesByVersion(releases))
+
+	var filteredReleases []Release
+
+	for i := 0; i < len(releases); i++ {
+		r := releases[i]
+
+		// Deepcopy changelogs as they are modified later and this instance of
+		// r ends up to filteredReleases.
+		r.changelogs = make([]Changelog, len(releases[i].changelogs))
+		copy(r.changelogs, releases[i].changelogs)
+
+		// Find previous release and map changelogs for quick lookup.
+		prevRelease := findPreviousRelease(r, releases[:i])
+		prevChangelogs := mapReleaseChangelogs(prevRelease)
+
+		// Process changelogs of current release removing ones present in
+		// previous release.
+		for _, clog := range r.Changelogs() {
+			_, exists := prevChangelogs[clog.String()]
+			if exists {
+				// r.Changelogs() returns a copy of changelogs so removal won't
+				// mess iteration in this case.
+				r.removeChangelogEntry(clog)
+			}
+
+		}
+
+		filteredReleases = append(filteredReleases, r)
+	}
+
+	return filteredReleases
+}
+
+// findPreviousRelease finds release that is older than argument r0. This
+// function expects that releases is sorted by version as it is iterated
+// backwards. If no previous release is found, empty Release is returned.
+func findPreviousRelease(r0 Release, releases []Release) Release {
+	r0Date, err := time.Parse(releaseTimestampFormat, r0.Timestamp())
+	if err != nil {
+		panic(err)
```

Comment:
Why are there still so many panics? I don't understand where we are using this, that makes panics okay.

Reply:
Panics here came from my thinking that in this case if parsing fails, it means timestamps are formatted wrong, which means `opsctl` validation broke and if we just pass error and later in the call chain log and ignore it, we don't necessarily notice the issue. When cluster-service panics, we definitely notice the issue but we still don't affect customer workloads.

I can change these to errors and there is a very strong point that library should never panic, but how can we make sure we make this case visible enough in cluster-service? Should we expose a metric of failed releases (I mean number of IndexReleases that couldn't be compiled to Release)?

Reply:
I think a metric for errors in cluster service should do the job. Paging oncall that way lets us add information to the alert which lets them react fast to what's happening.
The service panicking doesn't give oncall a lot information right away IMO, especially if it restarts and then starts crashlooping.

Reply:
> I think a metric for errors in cluster service should do the job. Paging oncall that way lets us add information to the alert which lets them react fast to what's happening.
The service panicking doesn't give oncall a lot of information right away IMO, especially if it restarts and then starts crashlooping.

Good point. It returns error now :)